### PR TITLE
[pytorch] bump variable version before redispatching

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#41522 [WIP] experiment reorder version bumping**
* #41269 [pytorch] bump up variable version regardless of differentiability

Relaxing version bumping to all ops is problematic, because composite op can
generate false bumping and break version check, e.g.:
```
Op A (x) {  // (e.g.: rrelu_)
  Op B (x) {  // (e.g.: rrelu_with_noise_)
    modify(x)
    bump(x)
    save(x)  // for some ops it calls `save(x.clone())` before `modify(x)`
             // those ops won't hit this problem
  }
  bump(x) *// wrong! false update causing x's version != saved x's version*
}
```

Reordering bumping can workaround false bumping for composite ops:
```
Op A (x) {  // (e.g.: rrelu_)
  bump(x)  // false bumping but harmless?

  Op B (x) {  // (e.g.: rrelu_with_noise_)
    bump(x)

    modify(x)
    save(x)
  }
}
```

This change seems to be semantic preserving, with a side effect that version
counter could be incremented by >1 per root op call.

Differential Revision: [D22567838](https://our.internmc.facebook.com/intern/diff/D22567838/)

Differential Revision: [D22567838](https://our.internmc.facebook.com/intern/diff/D22567838)

### CODE REVIEW

**Comment by albanD** [comment]
That sounds good from a high level point of view.
Does this mean that we can simplify the `save()` logic to always save when the grad_fn is created and remove the special case for inplace functions that details this to the end?


One issue you will have though is about the check inplace (only generated for differentiable functions).
```
Op A (x) {  // (e.g.: rrelu_)
  bump(x)  // false bumping but harmless?

  Op B (x) {  // (e.g.: rrelu_with_noise_)
    check_inplace(x)
    bump(x)

    modify(x)
    save(x)
  }
}
```
In this case, if the check inplace fails, you will have already bumped the version counter, leading to more issues.
We might want to make a similar change as the previous PR to the check inplace to always happen (even composite functions).

**Comment by ezyang** [comment]
The change is semantics preserving because of a broader invariant the system abides by today, which is that once you exclude Autograd from dispatch, there is guaranteed not to be any code that relies on values of version counters (so it does not matter where the bump happens within this segment.) But it sounds like we're on the road to break that invariant, which makes me nervous. I'm not sure I prefer this patch over a different version of this patch where version bumps only occur if Autograd is excluded.

**Comment by ljk53** [comment]
> That sounds good from a high level point of view.
> Does this mean that we can simplify the `save()` logic to always save when the grad_fn is created and remove the special case for inplace functions that details this to the end?
> 
> One issue you will have though is about the check inplace (only generated for differentiable functions).
> 
> ```
> Op A (x) {  // (e.g.: rrelu_)
>   bump(x)  // false bumping but harmless?
> 
>   Op B (x) {  // (e.g.: rrelu_with_noise_)
>     check_inplace(x)
>     bump(x)
> 
>     modify(x)
>     save(x)
>   }
> }
> ```
> 
> In this case, if the check inplace fails, you will have already bumped the version counter, leading to more issues.
> We might want to make a similar change as the previous PR to the check inplace to always happen (even composite functions).

@albanD - thanks for pointing this out. Nice catch! I agree we should do check_inplace() for composite op too to keep exactly the same semantics.

I have a (related but) more general question about the purpose of version bumping: Is it OK to bump the version if calling the OP can *possibly* update the value but not actually update on this particular invocation? Let's call it "false positive bumping". Before this PR I think "false positive bumping" can already happen to some extend, e.g.: if the modification is on some conditional branch in ATen C++, and/or the value is numerically unchanged for the given input params. But as long as the call can possibly update the value it's ok to bump and report error on backward pass. Is this understanding correct?

For this particular case, we are introducing a new "false positive bumping" when subsequent check_inplace() in Op B fails. Because check_inplace() fails loudly, users can fix the problem and recover from it if they are in Jupyter notebook. The real damage of this new "false positive bumping" is that they will need recreate the affected tensors to restore their version (which they don't have to do currently). Is my understanding about the damage of the new "false positive bumping" correct?

It reminds me that there are similar error-checking happening before increment_version(), e.g.:
```
  if (compute_requires_grad( out )) {
    throw_error_out_requires_grad("prod");
  }
```
Technically we need do this in composite ops, too?

**Comment by albanD** [comment]
>  Is it OK to bump the version if calling the OP can possibly update the value but not actually update on this particular invocation?

I don't think it is ok no.
The main reason being that we would falsely prevent people from calling backward on code that is valid (and was previously working).

Obviously, if you do things like `a.add_(0)`, the version will be bumped but that is expected.
But if the Tensor is untouched, it's version counter should not be bumped.

>  "false positive bumping" can already happen to some extend, e.g.: if the modification is on some conditional branch in ATen C++

I am not aware of any of these. If you have examples, we should fix them.

> Technically we need do this in composite ops, too?

Actually this might be a more general issue that would prevent the moving of the version bump.
Because the inner function can fail for any reason it wants. And we expect that it only do so before modifying things inplace.
So you wouldn't be able to move all the tests to the composite op :/

But I am open to discussing "Are we ok we bumping the version counter of a Tensor even though it was not used if the inner op raises an error. The inner op's error should be fixed anyway and will remove the bad version bump." ?

**Comment by ljk53** [comment]
> The change is semantics preserving because of a broader invariant the system abides by today, which is that once you exclude Autograd from dispatch, there is guaranteed not to be any code that relies on values of version counters (so it does not matter where the bump happens within this segment.) But it sounds like we're on the road to break that invariant, which makes me nervous. I'm not sure I prefer this patch over a different version of this patch where version bumps only occur if Autograd is excluded.

@ezyang Thanks for the feedback. Just want to clarify a few things:

1. "we're on the road to break that invariant" - Do you mean this PR breaks the invariant? or are you talking about the idea of moving version bumping outside Autograd (e.g. aten codegen) will break the invariant because non-Autograd code will start updating the values of version counters? or something else?

2. "relies on values of version counters" - do you mean by "read" the value or "write" the value or both?

3. "only occur if Autograd is excluded" - Just to clarify, do you mean we can use "if Autograd is excluded" to check whether it's on root op A (where Autograd is not excluded yet) or intermediate op B (where Autograd is excluded) so that we can bump only for intermediate ops?
But we are about to move bump to lower level kernels, e.g. aten codegen. So if we plan to set the exclude-autograd-guard in root ops' autograd kernel, then it needs to be set before redispatching to A's lower level kernel, right?
Similarly, if root op A doesn't call other ops, then we do need bump in A - so I feel it's hard to say "only bump on intermediate ops or root ops if they don't call intermediate ops". We can do the reverse, though - only bump at the root ops (if Autograd is included), to avoid bumping >1 times. But I feel it's different goal?

**Comment by ezyang** [comment]
Let me answer these in a different order:

> "relies on values of version counters" - do you mean by "read" the value or "write" the value or both?

Reads.

> "we're on the road to break that invariant"

Hmm, I had to think a bit to remember what I was worried about. I think it is something like this.

Imagine that you are looking at the execution of a program, and its straightline execution is alternating between autograd on and autograd off (autograd off happens when we go "below" autograd to do actual computation; real computation never happens in the autograd on universe). Whether or not we have modified a tensor is a property of any given "autograd off" segment, because we can assume that "autograd on" segments never do any real work. If an autograd off segment does a mutation, it is obligated to bump version counters. However, autograd off segments never actually *read* a version counter, so it doesn't matter if you do the bump at the beginning or the end. As your discussion with @albanD revealed, it is better to do it at the end, so that you don't bump the version counter if the operation errors. Little diagram:

```

            +--- version counter bump
AD          V AD               AD
XXXXXX--------XXXXXXXX---------XXXXXX
      Mutating        Pure
      No-AD           No-AD

```

What are we "on the road" to breaking? Well, the general concept behind this experiment is that we will unconditionally bump version counter even for composite ops. Composite ops don't turn autograd off! So now I can't really associate version counter bumps with no-AD blocks, because your picture looks like this

```
/----- span of composite --\

XXXX-----XXXXXXXXX-----XXXXX
^   ^
+---+- version counter bumps
```

The old invariant I had in my head no longer holds, we are in some new world order.

> "only occur if Autograd is excluded" - Just to clarify, do you mean we can use "if Autograd is excluded" to check whether it's on root op A (where Autograd is not excluded yet) or intermediate op B (where Autograd is excluded) so that we can bump only for intermediate ops?

I think this is right. Let me clarify my thinking some more.

The problem that you seem to be grappling with in this PR, is that if you "spray" version bumps everywhere schema implies there is a version bump, you end up with these spurious bumps from composites, as is the case in `rrelu_`. So let me posit the following invariants (riffing off of my discussion earlier):

1. Schema accurately says whether or not an operator, at some point, will mutate an input
2. The "actual" mutation only ever (i.e., where the dirty work is done) occurs when autograd is disabled

With these invariants in hand, I can infer that if a composite operator mutates, there must be *some* autograd-off region inside the composite where the dirty work happens. Then I can rest assured that THIS operator shall be responsible for bumping the version counter, and I (the autograd-on segment) am absolved of responsibility to do the bumping.

But why do this as a dynamic test? Well, my impression was that it was inconvenient for you to find out if an operator actually was composite or not (because this is ATen codegen, not tools codegen). The dynamic test sidesteps this problem so you can still "spray" the version bumps everywhere, using schema only.

> But we are about to move bump to lower level kernels, e.g. aten codegen. So if we plan to set the exclude-autograd-guard in root ops' autograd kernel, then it needs to be set before redispatching to A's lower level kernel, right?

Using the above discussion to answer this question: it depends if the root op is a composite op. If it is composite, then when it calls the lower level kernel, it will hit the VariableType implementation, which will exclude-autograd, and then you'll finally get to the lower level kernel, so everything is fine.

> Similarly, if root op A doesn't call other ops, then we do need bump in A - so I feel it's hard to say "only bump on intermediate ops or root ops if they don't call intermediate ops". We can do the reverse, though - only bump at the root ops (if Autograd is included), to avoid bumping >1 times. But I feel it's different goal?

If root op doesn't call any ops, it is not composite. Then root will have a VariableType implementation that excludes autograd before going on to the lower level.

**Comment by ljk53** [comment]
> > But we are about to move bump to lower level kernels, e.g. aten codegen. So if we plan to set the exclude-autograd-guard in root ops' autograd kernel, then it needs to be set before redispatching to A's lower level kernel, right?
> 
> Using the above discussion to answer this question: it depends if the root op is a composite op. If it is composite, then when it calls the lower level kernel, it will hit the VariableType implementation, which will exclude-autograd, and then you'll finally get to the lower level kernel, so everything is fine.

@ezyang thanks for your detailed explanation! I think the reason I was confused was because I forgot that composite op does NOT set exclude-autograd guard before lowering (calling its TypeDefault implementation). I somehow thought we would unconditionally set the exclude-autograd guard for all ops in VariableType.

```
Tensor conv2d(const Tensor & input, const Tensor & weight, const Tensor & bias, IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
  // NO guard!!!
  //So TypeDefault::conv2d can use this dynamic info to determine whether it should bump or not.
  auto result = TypeDefault::conv2d(input, weight, bias, stride, padding, dilation, groups);
  return result;
}
```

So looking at emit TypeDefault call: https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L1156-L1159
and bump_version: https://github.com/pytorch/pytorch/blob/master/tools/autograd/gen_variable_type.py#L1207-L1208
They both respect the same "strategy == 'use_derived'" condition, so it's guaranteed to be consistent.

We probably should simply delete these fallthrough VariableType kernels (strategy != 'use_derived') as @ailzhang talked about recently.


Here are two more follow up questions:

- If it relies on autograd kernels to set exclude-autograd bit, then can we still consolidate NoGradGuard and NonVariableTypeMode guard?
The reason we want to move out increment_version is that we want to disable autograd dispatch key if NoGradGuard is set. But now the increment_version still depends on autograd kernel...


- Can a non-composite op (strategy == 'use_derived') call another op (is it what you called "numerical composition"?) If so, the new approach might still have spurious bump, right?
Here is one example: (baddbmm_ calls addmm_: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L494)
```
aten::baddbmm_
at::CPUType::baddbmm_(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)
at::native::baddbmm__cpu(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar)
at::native::bmm_out_or_baddbmm_(at::Tensor&, at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar, bool) 
at::Tensor::addmm_(at::Tensor const&, at::Tensor const&, c10::Scalar, c10::Scalar) const 
aten::addmm_
```

baddbmm_ & addmm_ both increment version in VariableType:
https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/autograd/generated/VariableType_1.cpp.html#2037
https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/autograd/generated/VariableType_0.cpp.html#1757

Currently it would not call addmm_'s autograd kernel because it's already excluded-autograd in baddbmm_'s autograd kernel, so it won't bump version when calling addmm_.
But if we move increment_version out of addmm_'s autograd kernel, say, into CPUTYPE::addmm_, then it would bump again, even with exclude-autograd check.

**Comment by albanD** [comment]
>  then can we still consolidate NoGradGuard and NonVariableTypeMode guard?

I am not sure how this would work.
As of today, one means "do not create the graph" while the other means "ignore everything related to autograd".
The two main things in the middle here are the version counter bump and the view tracking.
And we need this distinction because we often don't want to create the graph but keep the two others to make sure we compute the correct gradients.
Or is the idea to make version counter and view tracking "correct at the Aten level" and that it could be used outside of the autograd?
Does that mean that we will need to have the composite idea in Aten to know which function should handle these or not?

**Comment by ljk53** [comment]
> Or is the idea to make version counter and view tracking "correct at the Aten level" and that it could be used outside of the autograd?

Yeah - I think the motivation is to move version counter and other stuff unrelated to "create the graph" out from VariableType (i.e. "correct at the ATen level").

> Does that mean that we will need to have the composite idea in Aten to know which function should handle these or not?

This is something I'm trying to avoid because currently ATen codegen doesn't have all the information about "composite" (i.e. "strategy == 'use_derived'" condition). And that's why I'm experimenting this PR and discuss other ideas to avoid relying on VariableType to decide whether to bump version or not.

But if there is not better option, I guess we need think about porting some autograd codegen bits into aten codegen...

**Comment by ljk53** [comment]
> Obviously, if you do things like `a.add_(0)`, the version will be bumped but that is expected.
> But if the Tensor is untouched, it's version counter should not be bumped.
> 
> > "false positive bumping" can already happen to some extend, e.g.: if the modification is on some conditional branch in ATen C++
> 
> I am not aware of any of these. If you have examples, we should fix them.

Right, I was thinking of both "numerically unchanged" ("a.add_(0)") and "conditionally unchanged" ("dropout_(a, p, train=false)"). I agree for the "conditionally unchanged" case spurious bump can be bad because it might falsely break valid code (the condition is deterministically true/false).

But dropout_ is an example against the claim "Schema accurately says whether or not an operator, at some point, will mutate an input". Besides the schema, we need check the value of `train`.

Currently dropout_(train=false) doesn't bump version because dropout_ happens to be a composite op and it defers bumping to the nested mul_ call. But I imagine we could write it in non-composite way, right?

**Comment by albanD** [comment]
> But I imagine we could write it in non-composite way, right?

The current codegen wouldn't handle it.
I don't have other example for version bump, but for views, there are a few case where we do this extra "jump" via a composite function to ensure this happens conditionally. For example, for `contiguous()`.

**Comment by ezyang** [comment]
> The reason we want to move out increment_version is that we want to disable autograd dispatch key if NoGradGuard is set. But now the increment_version still depends on autograd kernel...

It's fine right? There are two parts to the recipe:

* Move increment version into the actual CPU/CUDA kernels so they are always done even if we skip autograd. (By the way, now that I am thinking about this, it is actually really dangerous to try to do this without working shape checking metaprogramming, because backend extensions are basically guaranteed to get this wrong. So maybe we should actually have a VersionCounter dispatch key that handles bumping--and then you also get to turn off version counting in situations when you don't need it.)
* In my "YOLO it's too hard to figure out if I'm allowed to increment version or not in a location" idea, increment version happens when autograd is DISABLED (and we do NOT increment when it is enabled). 

> Can a non-composite op (strategy == 'use_derived') call another op (is it what you called "numerical composition"?) If so, the new approach might still have spurious bump, right?

That's right. In this case, I think that the spurious bump is harmless, because it's guaranteed that no one will observe (read) the intermediate version counter. But it is a little surprising!

----

Looking over the difficulties here, I think I have a new proposal: let's introduce a new VersionCounter (or maybe AutogradMeta) dispatch key which handles version counting. This key contains all autograd relevant stuff that needs to run unconditionally (even if `requires_grad=False`) and can be generated from `tools/`, bypassing the knowledge information you have. This leaves us with two guards (like we have now), but they are now factored to semantically make more sense:

* NoGradGuard is like `no_grad` and makes us skip constructing graphs on the inside.
* NoVersionCounter guard is a completely new guard, and it causes us to skip bumping version counters and constructing view information. This guard is one you would use in predictor, e.g., where you know that this useful will never get used by autograd a priori (whereas with `no_grad`, you might still need to track version information if you end up using it out of scope).

My primary concern with the new version counter is that we add another dynamic dispatch. I am kind of thinking now that we should just eat it (and if it really is expensive to do this, let's go optimize the dispatcher some more).
